crates/cadenza-gcode/src/parser.rs

```diff
@@ -0,0 +1,255 @@
+//! GCode parser implementation.
+
+use crate::{
+    ast::{Command, CommandCode, Line, Parameter, ParameterValue, Program},
+    error::{Error, Result},
+};
+
+/// Parse a GCode string into a Program.
+pub fn parse_gcode(input: &str) -> Result<Program> {
+    let mut lines = Vec::new();
+
+    for line_str in input.lines() {
+        lines.push(parse_line(line_str)?);
+    }
+
+    Ok(Program { lines })
+}
+
+/// Parse a single line of GCode.
+fn parse_line(line: &str) -> Result<Line> {
+    let line = line.trim();
+
+    // Handle empty lines
+    if line.is_empty() {
+        return Ok(Line::Empty);
+    }
+
+    // Handle comments (semicolon or parentheses)
+    if let Some(stripped) = line.strip_prefix(';') {
+        return Ok(Line::Comment(stripped.trim().to_string()));
+    }
+
+    // Handle inline comments - extract comment part
+    let (code_part, _comment) = if let Some(pos) = line.find(';') {
+        let (code, cmt) = line.split_at(pos);
+        (
+            code.trim(),
+            Some(cmt.strip_prefix(';').unwrap_or(cmt).trim()),
+        )
+    } else {
+        (line, None)
+    };
+
+    // If only comment remains after splitting, treat as comment
+    if code_part.is_empty() {
+        if let Some(cmt) = _comment {
+            return Ok(Line::Comment(cmt.to_string()));
+        }
+        return Ok(Line::Empty);
+    }
+
+    // Parse as command
+    let command = parse_command(code_part)?;
+
+    // For now, we ignore inline comments during parsing
+    // They could be preserved in the AST in a future enhancement
+    Ok(Line::Command(command))
+}
+
+/// Parse a GCode command.
+fn parse_command(input: &str) -> Result<Command> {
+    let input = input.trim();
+    if input.is_empty() {
+        return Err(Error::InvalidCommand("Empty command".to_string()));
+    }
+
+    // Split into tokens
+    let tokens: Vec<&str> = input.split_whitespace().collect();
```

Comment:
Why do we need to allocate for this? We should be able to incrementally iterate on the split_whitespace